src/shortcut.js

```diff
@@ -0,0 +1,330 @@
+  Blockly.ShortcutRegistry.registry.addKeyMapping(
+      metaC, Blockly.ShortcutItems.names.COPY);
+};
+
+
+/**
+ * Keyboard shortcut to copy and delete multiple selected blocks on
+ * ctrl+x, cmd+x, or alt+x.
+ */
+const registerCut = function() {
+  const cutShortcut = {
+    name: Blockly.ShortcutItems.names.CUT,
+    preconditionFn: function(workspace) {
+      const selected = Blockly.common.getSelected();
+      return !workspace.options.readOnly && !Blockly.Gesture.inProgress() &&
+      selected && selected.isDeletable() && selected.isMovable() &&
+          !selected.workspace.isFlyout;
+    },
+    callback: function() {
+      copyData.clear();
+      const apply = function(block) {
+        if (block &&
+            block.isDeletable() &&
+            !hasSelectedParent(block)) {
+          copyData.add(block.toCopyData());
+          if (block.workspace.isFlyout) {
+            return;
+          }
+          block.workspace.hideChaff();
+          if (block.outputConnection) {
+            block.dispose(false, true);
+          } else {
+            block.dispose(true, true);
+          }
+        }
+      };
+      const selected = Blockly.common.getSelected();
+      Blockly.Events.setGroup(true);
+      if (!blockSelection.size) {
+        apply(selected);
+      }
+      blockSelection.forEach(function(id) {
+        const block = Blockly.getMainWorkspace().getBlockById(id);
+        apply(block);
+      });
+      Blockly.Events.setGroup(false);
+      return true;
+    },
+  };
+
+  Blockly.ShortcutRegistry.registry.register(cutShortcut);
+
+  const ctrlX = Blockly.ShortcutRegistry.registry.createSerializedKey(
+      Blockly.utils.KeyCodes.X, [Blockly.utils.KeyCodes.CTRL]);
+  Blockly.ShortcutRegistry.registry.addKeyMapping(ctrlX, cutShortcut.name);
+
+  const altX =
+  Blockly.ShortcutRegistry.registry.createSerializedKey(
+      Blockly.utils.KeyCodes.X, [Blockly.utils.KeyCodes.ALT]);
+  Blockly.ShortcutRegistry.registry.addKeyMapping(altX, cutShortcut.name);
+
+  const metaX = Blockly.ShortcutRegistry.registry.createSerializedKey(
+      Blockly.utils.KeyCodes.X, [Blockly.utils.KeyCodes.META]);
+  Blockly.ShortcutRegistry.registry.addKeyMapping(metaX, cutShortcut.name);
+};
+
+/**
+ * Keyboard shortcut to paste multiple selected blocks on
+ * ctrl+v, cmd+v, or alt+v.
+ */
+const registerPaste = function() {
+  const pasteShortcut = {
+    name: Blockly.ShortcutItems.names.PASTE,
+    preconditionFn: function(workspace) {
+      return !workspace.options.readOnly && !Blockly.Gesture.inProgress();
+    },
+    callback: function() {
+      Blockly.Events.setGroup(true);
+      blockSelection.forEach(function(id) {
+        const block = Blockly.getMainWorkspace().getBlockById(id);
+        if (block) {
+          block.pathObject.updateSelected(false);
+        }
+      });
+      blockSelection.clear();
+      copyData.forEach(function(data) {
+        if (!data) {
+          return null;
```

Comment:
Is there any reason `data` would be false-y?

Reply:
Actually I also haven't encountered a situation that will falsify `data`, but I can see check exists in Blockly's paste here:
https://github.com/google/blockly/blob/61322294da70c770b32abb28b9d695af75f61a16/core/clipboard.js#L47-L49

Reply:
Ah whoops I was reading core wrong. I thought it accepted the copy data as a parameter, but it apparently does not.

Regardless, you're never adding anything falsey to the `copyData` set, so it should be fine to remove this check.

Reply:
Okay, then I'll remove this check.